### PR TITLE
[dev](perf): DSA pre-attention: plan-driven scorer dispatch, host-built CP layout metadata

### PR DESCRIPTION
**Pull request by jingqiny-99**
## Summary

This PR makes DSA indexer scoring correct for general packed THD inputs and removes repeated layout construction from the packed context-parallel hot path.

The central semantic correction is that packed sequence cardinality is independent of context parallelism: CP=1 can contain multiple sequences, and CP>1 can contain either one or many. Scorer selection now follows the actual THD metadata and kernel preconditions instead of using CP size as a proxy for sequence count.

The final branch preserves existing configuration and backend-hook compatibility. It does **not** claim full CUDA Graph support for packed DSA+CP. Such configurations remain constructible and receive an experimental-support warning; capture/replay support is follow-up work.

### 1. Plan-driven general THD scoring

- Add a total scoring-plan resolver with explicit outcomes for plain causal scoring, packed single-segment scoring, general packed THD scoring, unfused bounds, and decline.
- Run multi-sequence CP=1 scoring as one cuDNN THD call using the original packed Q/K/W tensors and cu-seqlens.
- Run CP>1 general packed scoring from the same THD metadata, using per-sequence zigzag query segments and gathered-key reorder metadata.
- Keep the single-sequence specialization only when its structural preconditions hold.
- Pass explicit key-position state through the production dispatcher so unsupported layouts reach the plan's `DECLINE` outcome before a cuDNN launch.
- Emit one diagnostic when a valid layout intentionally uses the existing unfused bounds path.

This removes both invalid implications: CP=1 no longer means “one sequence”, and CP>1 no longer means “multiple sequences”.

### 2. Host-built packed-CP layout metadata

- Record compact host Q/K cu-seqlens during the existing THD route-prebuild stage.
- Build all-rank query positions and gathered-key reorder tables from those host integers.
- Construct the reorder as a direct inverse permutation instead of a full argsort.
- Copy each finished table to the target device with a direct tensor transfer.
- Memoize layouts on the per-microbatch `PackedSeqParams`, so later DSA layers reuse the first build.
- Use the computed Q/K output-cover flags in both `positions_and_reorder` memo sites, so equal geometry has an equal cache key.

Independent CPU goldens cover CP1/2/4, both/all ranks, multiple sequences, zero-length entries, all-empty packs, padding, distinct Q/K boundaries, and every Q/K cover-flag combination. A production-forward regression test also enters both memo sites and proves that the second site reuses the first result.

### 3. Compatibility and invariant hardening

- Keep the existing backend-hook kwargs `use_local_indexer_varlen` and `single_packed_thd_sequence` on their CP-only contract.
- Offer CP-independent layout facts as separate optional kwargs.
- Move callable-signature filtering into `megatron.core.utils.filter_kwargs_for_callable`; exact signatures, `**kwargs`, positional-only parameters, opaque callables, and the identity-safe signature cache have direct tests.
- Preserve the pre-existing opaque full-hook `varlen_is_plain_causal` contract.
- Preserve valid self-attention Q/K metadata aliases through THD padding and inherited metadata.
- Reject inconsistent scorer devices, dtypes, views, TP slices, and gradient arity before producing incorrect results.
- Pass configured `cp_comm_type` into attention layers for Dynamic CP even when configured CP is 1; the effective microbatch CP group may still be larger than 1.
- Keep that effective Dynamic CP group inside selective core-attention checkpoint replay, then restore the caller's group after both the original call and recomputation.
- Restore the non-optional `starts`/`ends` autograd contract after the caller's concrete-bounds guard.

### 4. CUDA Graph scope

Packed DSA+CP graph configurations are no longer rejected by an empirical hardware/version matrix. The final check is structural and warning-only:

- static packed DSA warns when configured CP > 1 and CUDA graphs are enabled;
- Dynamic CP packed DSA warns whenever CUDA graphs are enabled, including configured CP=1, because the effective per-microbatch group can be larger;
- the warning is independent of scorer backend and capture module scope;
- eager execution (`cuda_graph_impl="none"`) is unchanged.

The evidence is diagnostic, not an allow/deny table. On the validation GB200 image, cuDNN static CP2 eager execution passes, while local attention capture reaches graph warmup and fails when the cloned packed metadata is empty. Backend `none` was previously measured only for static CP2/local/attention and also failed during graph warmup. TileLang was unavailable in that image and remains unmeasured. No unmeasured combination is presented as known-good, and none is hard-rejected by this PR.

Full capture/replay support for changing packed layouts is intentionally deferred.

## Validation

All GPU work below ran on Lyris GB200 in the standard `gb200` partition (no backfill), using `dsv4-gb200-torch2604-nvrx060`.

- Final consolidated targeted suite: **164 passed** — config compatibility/warnings, scoring plans, hook ABI, generic callable filtering, independent host-layout goldens and cache lifetime, cross-site memo reuse, Dynamic-CP communication and selective-recompute group lifetime, explicit-position dispatch, plus real cuDNN CP2 parity on both zigzag ranks and three seeds against an independent PyTorch forward/backward reference.
- Profiler-free model probes:
  - static CP2 eager: pass across one- and multi-sequence boundary changes;
  - Dynamic CP configured CP2 / effective CP1 eager: pass;
  - Dynamic CP configured CP1 / effective CP2 eager: pass after the communication-propagation fix;
  - static CP2 local-attention graph: configuration constructs and warns, eager preflight runs, then capture fails at graph warmup on empty cloned metadata (known deferred boundary).
- Selected H2D A/B: direct transfer reduced helper latency by **3.0%–17.9%** across the measured CP2/4/8/16 layouts versus pin-then-transfer, with bit-identical outputs. This is a helper-level measurement, not an end-to-end iteration percentage.

Only the conclusion-relevant H2D nsys report was retained; correctness probes were run without profiling.

## Commit map

1. Define explicit indexer scoring plans.
2. Preserve optional backend-hook compatibility.
3. Preserve THD Q/K metadata aliases.
4. Prebuild packed CP layout metadata.
5. Support general packed THD indexer scoring.
6. Add the initial packed DSA+CP graph check.
7. Preserve inherited THD metadata aliases.
8. Harden packed cuDNN scorer invariants.
9. Refine graph and hook compatibility from review feedback.
10. Final review follow-up: remove empirical hard rejection, fix memo/H2D/dispatcher contracts, generalize the hook helper, cover Dynamic CP configured-CP1 and selective recomputation, and expand independent/GB200 tests.

### CODE REVIEW

File: megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py [hunk 1]
```diff
@@ -415,6 +415,36 @@ def _causal_seq_lens(q_positions: Tensor, ratio: int, sk: int) -> Tensor:
...
+    branch, which reaches the fused single-kernel ``indexer_forward_wrapper``
+    instead of the per-head scoring loop in
+    :func:`_compute_indexer_scores_chunk_with_global_rows`.
+
+    The flag carries the mask builder's structural knowledge, so we avoid the
+    per-forward host/device sync that a ``torch.equal`` bounds comparison would
+    force on this common path. Genuine packed/CP/custom-position varlen is left
+    untouched (``varlen_is_plain_causal`` is False there).
+    """
+    if cp_size == 1 and packed_seq_params is None and varlen_is_plain_causal:
```
Comment (buptzyb) [inline megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py:490]: Can we let it reach the fused scorer when `packed_seq_params` is not None?

Comment (jingqiny-99) [reply megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py:490]: Yes. The current head routes packed THD independently of CP size: single-sequence packs use the packed-segment scorer and multi-sequence packs use the general THD scorer, both fused when their metadata preconditions hold. `_normalize_plain_causal_bounds` remains limited to the non-packed case because the packed paths must retain cu-seqlens instead of dropping their bounds. This is covered by the scoring-plan matrix for CP=1/2/16 and real CP=1 multi-sequence forward/backward parity against an independent PyTorch reference.

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 2]
```diff
@@ -2215,15 +2310,45 @@ def forward(
...
+                        (
+                            "positions_and_reorder",
+                            cp_size,
+                            cp_rank,
+                            local_len,
+                            local_len,
+                            local_len * cp_size,
+                            False,
+                            False,
+                        ),
```
[the remark this answers is not in the document]
Comment (jingqiny-99) [reply megatron/core/transformer/experimental_attention_variant/dsa.py:2324]: They are conflict branches, not memo-key mismatch.

File: megatron/core/transformer/experimental_attention_variant/dsa_scoring_plan.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+from enum import Enum
+from typing import Optional
+
+import torch
+
+
+class IndexerScoringPlan(Enum):
+    """How indexer scores and top-k are produced for one attention call."""
+
+    PLAIN_CAUSAL = "plain_causal"
```
Comment (buptzyb) [inline megatron/core/transformer/experimental_attention_variant/dsa_scoring_plan.py]: Why "PLAIN_xxx" instead of "UNPACKED_xxx", to align with the following "PACKED_xxx"?

File: megatron/core/transformer/experimental_attention_variant/dsa_scoring_plan.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    PACKED_SINGLE_SEGMENTS = "packed_single_segments"
+    """One sequence per pack, scored as one or more causal query/key segments."""
+
+    PACKED_THD_GENERAL = "packed_thd_general"
+    """General packed THD scoring driven by query/key cu_seqlens metadata."""
+
+    UNFUSED_BOUNDS = "unfused_bounds"
+    """No fused kernel claims this layout; scoring falls back to the per-head loop."""
+
+    DECLINE = "decline"
```
Comment (buptzyb) [inline megatron/core/transformer/experimental_attention_variant/dsa_scoring_plan.py:45]: Are we planning to eliminate unfsed bounds & decline cases gradually in the future?

File: megatron/core/transformer/experimental_attention_variant/dsa_kernels.py [hunk 5]
```diff
@@ -74,6 +76,74 @@ def _log_declined_hook(config: TransformerConfig, hook_name: str, reason: str) -
     )
 
 
+def _hook_kwargs_accepting(fn, *, opaque_fallback_names: Tuple[str, ...] = (), **candidate_kwargs):
```
Comment (buptzyb) [inline megatron/core/transformer/experimental_attention_variant/dsa_kernels.py]: I think this is a very general helper. Should we move this to some files like utils.py?

File: megatron/core/transformer/experimental_attention_variant/dsa_layout.py [hunk 6]
```diff
@@ -357,6 +452,161 @@ def build_packed_allgather_cp_query_positions_and_key_reorder(
...
+            )
+        half = seq_len // cp_size // 2
+        if half <= 0:
+            continue
+        spans.append((seq_start + cp_rank * half, half))
+        spans.append((seq_end - (cp_rank + 1) * half, half))
+    return spans
+
+
+def _host_to_device(values: torch.Tensor, device: torch.device) -> torch.Tensor:
```
Comment (buptzyb) [inline megatron/core/transformer/experimental_attention_variant/dsa_layout.py:486]: Don't we even have a general h2d/d2h helper function in mcore as of 2026.9?

File: megatron/core/transformer/cuda_graph_config.py [hunk 7]
```diff
@@ -59,6 +59,110 @@ def validate_moe_cuda_graph_support(config) -> None:
...
+
+            if modules not in measured_scopes:
+                return False
+            if not modules:
+                # ``block`` owns an inference-only TransformerBlock graph and
+                # skips the per-layer whole-scope manager exercised by the
+                # training validation matrix.
+                return inference_cuda_graph_scope == InferenceCudaGraphScope.layer
+            return True
+        if cuda_graph_impl == "transformer_engine":
```
Comment (buptzyb) [inline megatron/core/transformer/cuda_graph_config.py]: elif

File: megatron/core/transformer/cuda_graph_config.py [hunk 8]
```diff
@@ -59,6 +59,110 @@ def validate_moe_cuda_graph_support(config) -> None:
...
+    )
+    if experimental_attention_variant != "dsa":
+        return False
+    if not (static_packing or dynamic_packing):
+        return False
+
+    # Capture scope execution uses membership checks, so repeated spellings such
+    # as ``attn,attn`` are the same effective scope as ``attn``.
+    modules = frozenset(cuda_graph_modules or [])
+    if dsa_kernel_backend == "cudnn":
```
Comment (buptzyb) [inline megatron/core/transformer/cuda_graph_config.py]: Are tilelang/none backends verified supported, or just not considered?

File: megatron/core/transformer/experimental_attention_variant/dsa_scoring_plan.py (hunk 4 above)
Comment (jingqiny-99) [reply megatron/core/transformer/experimental_attention_variant/dsa_scoring_plan.py:45]: In my opinion, yes.

File: megatron/core/transformer/experimental_attention_variant/dsa_kernels.py (hunk 5 above)
Comment (jingqiny-99) [reply megatron/core/transformer/experimental_attention_variant/dsa_kernels.py]: Agree, and I'll move it immediately.

File: megatron/core/transformer/cuda_graph_config.py (hunk 7 above)
Comment (jingqiny-99) [reply megatron/core/transformer/cuda_graph_config.py]: OK!

File: megatron/core/transformer/cuda_graph_config.py (hunk 8 above)
Comment (jingqiny-99) [reply megatron/core/transformer/cuda_graph_config.py]: Tilelang and none have not been tested before, and I can test them now.

File: megatron/core/transformer/experimental_attention_variant/dsa_layout.py (hunk 6 above)
Comment (jingqiny-99) [reply megatron/core/transformer/experimental_attention_variant/dsa_layout.py:486]: I'm not sure, but I tried to find similar helpers by matching the name or asking agents like Claude or Codex, and found nothing T_T

Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/dsa_layout.py:486]: So, may we also move this helper to somewhere general?

File: megatron/core/transformer/experimental_attention_variant/dsa_scoring_plan.py (hunk 4 above)
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/dsa_scoring_plan.py:45]: So should we document what known cases would go to unfused bounds and decline for now?

File: megatron/core/transformer/experimental_attention_variant/dsa_layout.py (hunk 6 above)
Comment (jingqiny-99) [reply megatron/core/transformer/experimental_attention_variant/dsa_layout.py]: Agree. I will move it to a general place.

File: megatron/core/transformer/experimental_attention_variant/dsa_scoring_plan.py (hunk 4 above)
Comment (jingqiny-99) [reply megatron/core/transformer/experimental_attention_variant/dsa_scoring_plan.py:51]: OK! I'll summarize and document these cases.

File: megatron/core/utils.py [hunk 9]
```diff
@@ -691,6 +762,11 @@ def get_model_config(model):
     return get_attr_wrapped_model(model, "config", allow_none=False)
 
 
+def move_host_tensor_to_device(values: torch.Tensor, device: torch.device) -> torch.Tensor:
```
Comment (buptzyb) [inline megatron/core/utils.py]: Should we merge this helper with `maybe_move_tensor_to_cpu`? https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/moe/moe_utils.py#L1238

I think `maybe_move_tensor_to_cpu` is in a bad place. Moving it here would be better.

BTW looks like `maybe_move_tensor_to_cpu` is using `non_blocking` always. We may need a new argument `synchronization: bool = True`, which determines whether we add a stream sync after the non-blocking copy.

Ignore me if you think this is a bad idea.

File: megatron/core/transformer/transformer_layer.py [hunk 10]
```diff
@@ -371,7 +371,9 @@ def __init__(
         )
 
         attention_optional_kwargs = {}
-        if config.context_parallel_size > 1 and config.cp_comm_type is not None:
+        if (
+            config.context_parallel_size > 1 or config.dynamic_context_parallel
```
Comment (buptzyb) [inline megatron/core/transformer/transformer_layer.py:375]: I saw somewhere in `transformer_config.py` still reads `if config.context_parallel_size > 1: xxxx`. Should we align them to `if config.context_parallel_size > 1 or config.dynamic_context_parallel: xxxx` as well?

File: megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py [hunk 11]
```diff
@@ -2130,8 +2504,8 @@ def forward(
         sparse_loss: bool,
         calculate_per_token_loss: bool,
         d_v: int,
-        varlen_starts: Optional[Tensor],
-        varlen_ends: Optional[Tensor],
+        varlen_starts: Tensor,
```
Comment (buptzyb) [inline megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py]: Not optional anymore?

Comment (jingqiny-99) [reply megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py]: Sorry for this mistake... I have reverted this change

File: megatron/core/transformer/transformer_layer.py (hunk 10 above)
Comment (jingqiny-99) [reply megatron/core/transformer/transformer_layer.py:375]: Agree, I updated other checks to also cover dynamic_context_parallel.

File: megatron/core/utils.py (hunk 9 above)
Comment (jingqiny-99) [reply megatron/core/utils.py]: I think you are right, the `maybe_move_tensor_to_cpu` is in a strange place, I think it's good to move it here, and merge them in the future, because in my opinion, it may become a large helper (move tensor to arbitrary device and maybe do some synchronization if need), and it isn't transparent to its callers.

File: megatron/core/utils.py [hunk 12]
```diff
@@ -691,6 +762,36 @@ def get_model_config(model):
...
+    """Move a host tensor to ``device``."""
+    return values.to(device, non_blocking=device.type == "cuda")
```
Comment (buptzyb) [inline megatron/core/utils.py:782]: Do we need `values.pin_memory().to(device)` to avoid pageable H2D?

Comment (jingqiny-99) [reply megatron/core/utils.py:782]: I think it is right, but in my test case, removing `pin_memory` will save copy time. So I will add an argument here to decide whether run `values.pin_memory()` here, and its default value will be False.

Comment (buptzyb) [reply megatron/core/utils.py:782]: Is it because your `values` are already on pin? Non-pin h2d is dangerous since pagable copy is blocking and non-graphable IIRC.

Comment (jingqiny-99) [reply megatron/core/utils.py:782]: soga..Thanks for teaching me, and sorry for my limited knowledge. I will pin the memory by default.

Comment (buptzyb) [reply megatron/core/utils.py:782]: Happen to read an article about a hang issue caused by pagable memcpy: https://mp.weixin.qq.com/s/uOJALDKb588mFynaeJQJoA

**Review by buptzyb — APPROVED** [review]
I have to admit I cannot understand every single detail of this PR. But the code organization and functionality generally LGTM. One concern is the main branch is also making their refactoring to the dsa code (like #7211), so the main-dev divergency would be larger and larger. Beg @hxbai for a final review, thanks!

**Review by yaox12 — APPROVED** [review]
